src/Symfony/Component/Security/Guard/Provider/GuardAuthenticationProvider.php

```diff
@@ -140,8 +139,31 @@ private function authenticateViaGuard($guardAuthenticator, PreAuthenticationGuar
         return $authenticatedToken;
     }
 
+    private function findOriginatingAuthenticator(PreAuthenticationGuardToken $token)
+    {
+        // find the *one* GuardAuthenticator that this token originated from
+        foreach ($this->guardAuthenticators as $key => $guardAuthenticator) {
+            // get a key that's unique to *this* guard authenticator
+            // this MUST be the same as GuardAuthenticationListener
+            $uniqueGuardKey = $this->providerKey.'_'.$key;
+
+            if ($uniqueGuardKey === $token->getGuardProviderKey()) {
+                return $guardAuthenticator;
+            }
+        }
+
+        // no matching authenticator found - but there will be multiple GuardAuthenticationProvider
+        // instances that will be checked if you have multiple firewalls.
+
+        return null;
```

Comment:
does Symfony use explicit null returns at the end of function bodies? (I'm not sure, so don't immediately change this)

Reply:
yes we do :) this method could have a `?AuthenticatorInterface` return type starting from 4.0, explicit null would be mandatory

Reply:
Those two are mutually exclusive, a nullable return type cannot have an implicit return value of null. Should I fix something?

Reply:
@biomedia-thomas nope. If this patch gets merged, it must be on 2.8 (the lowest maintained branch where it applies) where return types cannot be added due to support of old PHP versions. We'll take care of adding it when merging it up to master